Added `MasonryFlashList` which adds support for rendering masonry layouts

naqvitalha opened the pull request:
## Description

resolves #455 
resolves #575 
resolves #589 

This PR adds support for Masonry layout rendering also referred to as Staggered Grid View layout. 

It's implemented using core FlashList and I didn't plan to make it part of this repo (initially) because it's technically like a derivative. I changed my mind because at the end of the day masonry is a layout and how it's implemented shouldn't matter. Ideally, anything that can be built using FlashList can be kept out. Keeping it lean is a win and should keep the core quality high.

I took the liberty to fix two additional bugs I found while building this otherwise there's no change in core. I've added comments to those changes.

### Why use this approach?

It can be implemented using nested lists without any drawbacks and also doesn't require a new layout algorithm on the native side. It's incredibly fast as well.

### Performance

It's incredibly well performing. Compared to 2x2 non masonry this is just 2% slower which may very well be a measurement problem.

### Known issue

There's a prop type warning being thrown by RLV which I'll address when we upgrade RLV the next time.

## Reviewers’ hat-rack :tophat:

- [ ] Open fixture and try the masonry sample

## Screenshots or videos (if needed)

<img src="https://user-images.githubusercontent.com/7811728/188957840-687810a1-5f44-43de-a465-859403b210ca.png" width="400">

## Checklist

- [x] I have added a changelog entry following the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) guidelines.

naqvitalha commented:
> This looks really good 👏
> 
> I have a couple of comments and questions but overall, I love the idea of having multiple `FlashList`s ❤️
> 
> Additionally, I got this warning when running the `Masonry` example: ![image](https://user-images.githubusercontent.com/9371695/190129862-9b3c4659-cc83-4482-968a-b19d7e3e27f1.png)

This requires a proptype fix in RLV. I'll get to it later

naqvitalha commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+## Methods
+
+`MasonryFlashList` exposes the some methods that `FlashList` does. These are:
```
We only have 3 while FlashList has many

fortmarek replied:
ahh, ok, have not realized 👍

naqvitalha commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+## Methods
+
+`MasonryFlashList` exposes the some methods that `FlashList` does. These are:
```
FlashList has many more methods which aren't going to work right now.

naqvitalha commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+    );
+
+    const onScrollRef = useRef<OnScrollCallback[]>([]);
+    const emptyScrollEvent = useRef(getBlackScrollEvent())
+      .current as NativeSyntheticEvent<MasonryFlashListScrollEvent>;
```
It should be fine. Those are the only fields we depend on. The event users listen to will have full data.
If we change anything the unit tests will cathc it.

naqvitalha commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+        emptyScrollEvent.nativeEvent.doNotPropagate = true;
+        onScrollProxy?.(emptyScrollEvent);
+        emptyScrollEvent.nativeEvent.doNotPropagate = false;
+      }, 32);
```
This onScroll doesn't scroll or anything. It just calls internal methods and updates offsets so that all viewability events are accurate. 32 is two frames which is the minimum I need.
Internal list offsets can be wrong if you add a header to the top list. I'll add this to the comments

fortmarek replied:
> 32 is two frames which is the minimum I need.

let's also add information about this. The amount of time for two frames also depends on the screen refresh rate but I'm ok with treating 60 Hz as the default.

naqvitalha commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+      CellRendererComponent,
+      ItemSeparatorComponent,
+      ...remainingProps
+    } = props;
```
Nothing specific. Having closer to render makes it easier to keep a track of things being filtered.

fortmarek commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+    );
+
+    const onScrollRef = useRef<OnScrollCallback[]>([]);
+    const emptyScrollEvent = useRef(getBlackScrollEvent())
```
It takes just one line, max two. But I'm ok either way, definitely minor.

naqvitalha commented on documentation/docs/guides/masonry-layout.md:
````diff
@@ -0,0 +1,85 @@
+```tsx
+// if `numColumns` is 3, you can return 1.25 for index 1 and 0.75 for the rest to achieve a 1:2:1 split by width.
+getColumnSizeMultiplier={(items, index, maxColumns, extraData) => {
+    return index === 1 ? 0.75 * 2 : 0.75 * 1;
````
Done 👍

naqvitalha commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+    }
+    for (let i = 0; i < dataSize; i++) {
+      let nextColumnIndex = i % columnCount;
+      if (autoOptimizeLayout) {
```
While I agree in general my main concern is with implicitly changing the order of items. I referred to Android's StaggeredGridLayoutManager and it keeps optimize on by default. The difference is that it makes sure that you, visually, see the optimization happen and we can't do that here. Animations across columns aren't possible in the current design. Making it apparent is a theme, even SwiftUI libraries do that same.

My worry is that by doing it implicitly we will confuse people as other open source packages in RN don't do this at all. Folks debugging the data order is one possible side effect.

Also, overrideItemLayout isn't trivial to implement. It increases the entry barrier by a lot (almost as much as RLV). It might be better to let people turn it on if they need it. I also wonder if this is a real concern when pagination is involved.

I can see the point but reorder not being apparent is a problem. Let me think about this more otherwise we can just wait to see how people use it and make the change.

I had also considered a new prop for this. Something simpler to implement but introducing one more thing to learn didn't seem very ideal.

naqvitalha commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,409 @@
+      setTimeout(() => {
+        emptyScrollEvent.nativeEvent.doNotPropagate = true;
+        onScrollProxy?.(emptyScrollEvent);
+        emptyScrollEvent.nativeEvent.doNotPropagate = false;
```
The native scrollEvents will not set this value. That's why I kept something that'll have false value by default.

Slapbox commented:
How are the odds this will be merged?

naqvitalha commented:
@fortmarek this is good to go IMO. Thoughts?

fortmarek commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+<img src="https://user-images.githubusercontent.com/7811728/188055598-41f5c961-0dd0-4bb9-bc6e-22d78596a036.png" height="500"/>
+</div>
+
+To get started, import `MasonryFlashList` from `@shopify/flash-list` and use it just like you would use `FlashList`:
```
Might be worth here just putting a code snippet? I know it will be basically identical with `flash-list` but imho still a good practice.

fortmarek commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+## Unsupported Props
+
+There are some props that `MasonryFlashList` does not support when compared to `FlashList`. These are:
```
```suggestion
There are some props that `MasonryFlashList` does not support when compared to `FlashList`:
```

fortmarek commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+## Additional Props
+
+There are some additional props that `MasonryFlashList` supports when compared to `FlashList`. These are:
```
```suggestion
`MasonryFlashList` supports these additional props on top of `FlashList`:
```

`when compared to` sounds odd to me here and we can make the wording more concise.

fortmarek commented on documentation/docs/guides/masonry-layout.md:
````diff
@@ -0,0 +1,85 @@
+### `getColumnFlex`
+
+```tsx
+  getColumnFlex?: (
+    items: T[],
+    columnIndex: number,
+    maxColumns: number,
+    extraData?: any
+  ) => number;
````
```suggestion
getColumnFlex?: (
  items: T[],
  columnIndex: number,
  maxColumns: number,
  extraData?: any
) => number;
```

There's extra padding for this prop which there should not be afaict

fortmarek commented on documentation/docs/guides/masonry-layout.md:
````diff
@@ -0,0 +1,85 @@
+optimizeItemArrangement?: boolean;
+```
+
+If enabled, MasonryFlashList will try to reduce difference in column height by modifying item order. `overrideItemLayout` is required to make this work. Default value is `false`.
````
```suggestion
If enabled, MasonryFlashList will try to reduce difference in column height by modifying item order. If `true`, specifying [`overrideItemLayout`](../fundamentals/usage.md#overrideitemlayout) is required. Default value is `false`.
```

fortmarek commented on documentation/docs/guides/masonry-layout.md:
````diff
@@ -0,0 +1,85 @@
+  ) => number;
+```
+
+Allows you to change the column widths of the list. This is helpful if you want some columns to be wider than the others.
````
```suggestion
`getColumnFlex` allows you to change the column widths of the list. This is helpful if you want some columns to be wider than the others.
```

fortmarek commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+Scroll to a specific content pixel offset in the list.
+
+Param `offset` expects the offset to scroll to.
```
```suggestion
Parameter `offset` expects the offset to scroll to.
```

fortmarek commented on documentation/docs/guides/masonry-layout.md:
```diff
@@ -0,0 +1,85 @@
+
+Param `offset` expects the offset to scroll to.
+
+Param `animated` (`false` by default) defines whether the list should do an animation while scrolling.
```
```suggestion
Parameter `animated` (`false` by default) defines whether the list should animate while scrolling.
```

fortmarek commented on fixture/src/Masonry.tsx:
```diff
@@ -0,0 +1,112 @@
+  );
+}
+
+const Component = (props: {
+  item: MasonryData;
+  text?: string;
+  backgroundColor?: string;
+}) => {
```
We should still treat it with care, though. I'd personally change that but it's minor.

fortmarek commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,438 @@
+    ).current;
+
+    /**
+     * We're triggering an onScroll on internal lists so that they register the correct offset which is offset - header size.
```
```suggestion
     * We're triggering an onScroll on internal lists so that they register the correct offset which is equal to offset - header size.
```

fortmarek commented on src/MasonryFlashList.tsx:
```diff
@@ -0,0 +1,438 @@
+
+/**
+ * FlashList variant that enables rendering of masonry layouts.
+ * `MasonryFlashList` will not compute most optimal arrangement for you automatically. You can use the `useMasonryLayoutOptimizer` preprocessing hook to improve item arrangement.
```
`useMasonryLayoutOptimizer` does not seem to exist? 🤔

fortmarek commented:
I have only a couple of minor comments, mostly about the documentation.

Additionally, as mentioned [here](https://github.com/Shopify/flash-list/pull/587#discussion_r976229222), for now I'd mark this component as unstable or alpha before we are more confident in the stability of this feature.

naqvitalha commented:
@fortmarek Addressed the comments! Thanks for the review again.
Regarding stability, it's very stable and ready for prime time. I don't see a need for marking it alpha or beta. The warning that you see is a proptype warning and doesn't impact stability in any way.

fortmarek reviewed (approved):
Let's goo 🚢 🚢

farhanhaider1 commented:
> > This looks really good 👏
> > I have a couple of comments and questions but overall, I love the idea of having multiple `FlashList`s ❤️
> > Additionally, I got this warning when running the `Masonry` example: ![image](https://user-images.githubusercontent.com/9371695/190129862-9b3c4659-cc83-4482-968a-b19d7e3e27f1.png)
> 
> This requires a proptype fix in RLV. I'll get to it later

any update on this?
i am still getting this warning.

naqvitalha commented:
It should be fixed in v1.4.0

fortmarek commented:
> Does overrideItemLayout not work with Masonry layout? If so, is that documented anywhere? We can't get overrideItemLayout working at all.

Can you flag an issue for this, please? 🙏 I'm personally not aware that Masonry layout doesn't support `overrideItemLayout`.

naqvitalha commented:
Yes, it's not possible to change spans in masonry. We'd appreciate a PR updating the docs!

Slapbox commented:
That's terribly upsetting to hear! We just did hours of work to implement `MasonryFlashList`. Isn't there some way to allow `overrideItemLayout` - even if it's imperfect? I'd much rather that to having to roll back hours of work - considering `FlashList` appears to be the future and I'm confident a good/proper solution will exist in the future even if it doesn't today.

Any chance to simply patch masonry to accept `overrideItemLayout`?

Slapbox commented:
@naqvitalha if this is treated as a docs-only issue - at minimum these pages need updating:

- https://shopify.github.io/flash-list/docs/guides/section-list 
    - "but all of them are replaceable with existing props."
    - `renderSectionHeader` doesn't seem to work with masonry
    - Surely other properties are misleadingly treated as though they would work with masonry
- https://shopify.github.io/flash-list/docs/guides/masonry
